### DISCUSSION TITLE
Async when decorator?

### DISCUSSION THREAD

**Discussion by Dharengo** [pyscript/pyscript#2156] [Proposals]
Is it a good QoL idea to add an async version of the when decorator? It would allow one to decorate async functions the same way they would decorate normal ones.

**Comment by WebReflection** [comment]
@JeffersGlass any thoughts around this? I don't see "*why not*", personally ... btw, we enabled async top-level await by default recently, I am not even sure that works already with the current state.

**Reply by Dharengo** [reply]
I would assume that doesn't magically allow us to await from within a non-async function though?

**Reply by WebReflection** [reply]
why not?

```python
@when("click", "#my_button")
async def click(event):
  print(event.type)
  # await something(...)
```

or better, why shouldn't it?

**Reply by WebReflection** [reply]
to clarify, I am not stating this works now, I am just wondering *why shouldn't it* as I see it pretty natural to write and deal with

**Reply by WebReflection** [reply]
truth to be told, async listeners are a feature and a curse ... if you expect, as example, `event.preventDefault()` to happen *after* anything asynchronous that'd be too late for the DOM to deal with your intention ... it's a long standing thing on the Web that only recently received enough attention but I am afraid there's no ETA around when, or if ever, that would land.

**Reply by Dharengo** [reply]
> truth to be told, async listeners are a feature and a curse ... if you expect, as example, `event.preventDefault()` to happen _after_ anything asynchronous that'd be too late for the DOM to deal with your intention ... it's a long standing thing on the Web that only recently received enough attention but I am afraid there's no ETA around when, or if ever, that would land.

I already thought of the preventDefault() use case. Unfortunately, an async @ when might not be able to cleanly prevent the default from happening (because I assume the default could possibly happen before the async function is run). Unless you maybe add a "preventDefault" flag to the argument list for @ when().

Still, allowing us to use @ when on async functions seems like a neat idea.

>why not?

Because normally you can't use await from within a non-async function. I do not think a top-level await changes this rule.

```py
def non_async():
  await some_coroutine()
```

Would still raise an error I'd think.

**Reply by WebReflection** [reply]
I am not sure I am following ... I am talking about:

```python
@when("click", "#my_button")
async def click(event):
  print(event.type)
  # await something(...)
```

I am not talking about:

```python
async @when("click", "#my_button")
def click(event):
  print(event.type)
  # await something(...)
```

and I don't even know if that's a thing (it shouldn't?) ... so, in the first case, if we are aligned, the only difference is that the decorator adds an async function as listener, which is well allowed on the DOM and quite common these days.

In that regard, I don't see why it shouldn't be possible ... if it doesn't work already, it's about wrapping a task into a *JS* friendly *async* function, and that should be relatively trivial.

> Unless you maybe add a "preventDefault" flag to the argument list for @ when().

That's the solution already in our stack: if you add an `invoke` field to the extra 3rd argument of `addEventListener` you can pass a string or an array of strings such as `['preventDefault', 'stopPropagation']` and both methods will be automatically invoked on the event side synchronously, whenever that happens ... a similar mechanism could be used for `@when` also because `@when` doesn't accept the 3rd argument which is indeed very helpful for some listener or behavior, most relevant being: `{ "once": True }` when it comes to automatic removal of the event once it happens.

As you can see there is room for improvements, and adding both a 3rd argument + handling async tasks out of the box, seem to be a good improvement to me, not sure what others think though.

**Reply by Dharengo** [reply]
So if adding the third argument is as simple as... Just using the third argument that addEventListener already has... All that's left is to wrap the coroutine in a function that returns something that javascript recognizes as a Promise-like object, no? Either that, or wrap it in a function that invokes `asyncio.create_task()`. That might be easier?

**Comment by WebReflection** [comment]
So ... I've just tried this page and everything works already in both *Pyodide* and *MicroPython* ... it doesn't cover the option for `once` or other things but it means that the original discussion is already satisfied by our project.

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@pyscript/core/dist/core.css">
    <script type="module" src="https://cdn.jsdelivr.net/npm/@pyscript/core/dist/core.js"></script>
</head>
<body>
  <script type="mpy">
    from pyscript import window, document, fetch, when

    @when('click', '#click')
    async def click(event):
      text = await fetch(window.location.href).text()
      document.getElementById('output').append(text)
  </script>
  <button id="click">click</button>
  <pre id="output"></pre>
</body>
</html>
```

I wonder if this was even tested before discussing the feature but I am not sure this is what the feature request was about ... if that's the case, I think we can resolve this discussion.

/cc @ntoll and @mchilvers just as *FYI*

**Reply by WebReflection** [reply]
P.S. I have also added the very same test case https://github.com/pyscript/pyscript/pull/2157

**Reply by Dharengo** [reply]
I swear I tried this before but it didn't work.

**Reply by WebReflection** [reply]
not sure that was the case but something I keep failing myself to remember is: "*is the first the selector or the listener name?*" ... if you swap those two nothing happens so maybe using `async def click(event_type="click", selector="#click"):` would be a better approach, although I am not sure we called that `event_type` instead of just `type` but ... hey ... explicit is usually better than implicit.

**Reply by Dharengo** [reply]
I literally just tried it right now: 

`/home/pyodide/pyscript/event_handling.py:45: RuntimeWarning: coroutine '_init.<locals>.async_callback' was never awaited`

**Reply by Dharengo** [reply]
Ok, I figured it out. The reason it did not work for me is because for async function it _only_ works if you do include the 'event' argument in the callback.

Non-async callbacks allow for zero argument signatures. That's why I didn't think much of it and didn't attempt to try the 1-argument one before now. The error also made no mention of it lacking an argument in its signature, only that the function wasn't awaited.

**Reply by WebReflection** [reply]
every. single. listener. in the DOM receives one and only one argument which is the `event` ... since about ever ... and that's an *EventTarget* interface every events based thing implement too. ~~The dance around checking the signature cannot really work with tasks~~ (amend, we were missing an extra check for coroutines) so I am afraid ~~I don't see any easy way to fix that~~ (amend, it's fixed in here https://github.com/pyscript/pyscript/pull/2158), except that when you add a listener, `*args` or at least a single `event` (or `_` in JS world when a generic reference doesn't matter) is on you, for the time being ... but I don't think we should overly complicate the code to fix async tasks as listeners and I am actually happy that everything really works as expected.

**Reply by Dharengo** [reply]
You asked why I made this proposal in the first place. I am merely explaining to you the events that lead to me making this proposal. For synchronous functions, zero argument callbacks are allowed, for asynchronous ones, they are not, but I did not know this.

If the error says "callback was never awaited", how was I supposed to know that the problem was with the function signature?

Either way, the mystery is now solved.

Still a proponent of giving a third argument to @ when though. Which falls outside of the scope of the original proposal, but I still think it's a good idea.

**Reply by WebReflection** [reply]
by no mean I meant to come back badly or blame you for anything ... it's just a classic example where there is no bear minimum example I could test and think about, so here I just created my own example which worked out of the box because I know that events in DOM always receive an `event` and trust me when I say to build this project I failed at remembering that myself every single time and learned to add `*args` everywhere "*just because*" ... in my reply I was trying to provide details so that you won't fail easily the way I do too but I'll try to investigate if there is a reason for the signature introspection to fail on tasks and not on functions but again, if anyone adds a listener, they are better off adding an event too by default ... I hope this answer makes sense.

I'll check the rest and if not too complicated, I'll try to land a fix for the error.

**Reply by WebReflection** [reply]
P.S. there is also a difference with Pyodide VS MicroPython as latest doesn't have any signature introspection so *maybe* the gotcha was in the interpreter? I'll try to figure out but I'd love to know if you tried and failed with both Pyodide and MicroPython, thanks.

**Reply by Dharengo** [reply]
My entire application will break if I use mpy, so I haven't.

**Reply by WebReflection** [reply]
then I think this dance is overkill and not producing the desired outcome ... the MicroPython callback seems to indeed not suffer the mentioned issue: https://github.com/pyscript/pyscript/blob/main/pyscript.core/src/stdlib/pyscript/event_handling.py#L39-L61 ... maybe I can just use MicroPython code in there and call it a day ...

**Reply by WebReflection** [reply]
aha ... fixed in here, it's all good to me there and the fix easy enough ... the MicroPython thing is different and it resolves anyway, I had to manually check for a coroutine and resolve accordingly: https://github.com/pyscript/pyscript/pull/2158

**Reply by Dharengo** [reply]
Even though the nature of the problem I ran into was entirely different from what I initially assumed it was, I'm glad that me bringing this up has resulted in something.

**Comment by WebReflection** [comment]
the PyScript preview that's going to be released is:

  * **JS** https://cdn.jsdelivr.net/npm/@pyscript/core@0.5.9/dist/core.js
  * **CSS** https://cdn.jsdelivr.net/npm/@pyscript/core@0.5.9/dist/core.css

Everything seems fine including `async` when events.
